### PR TITLE
[BEAM-646] Provide local tags in PInput, POutput expansions

### PR DESCRIPTION
**Pull request by tgroh**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

Output an ordered colleciton in PInput and POutput expansions.

This provides information that is necessary to reconstruct a PInput
or POutput from its expansion.

### CODE REVIEW

**Comment by tgroh** [comment]
R: @kennknowles

File: sdks/java/core/src/test/java/org/apache/beam/sdk/values/PCollectionListTest.java [hunk 1]
```diff
@@ -44,4 +54,68 @@ public void testEmptyListFailure() {
...
+
+    List<TaggedPValue> expansion = fromEmpty.expand();
+    // TaggedPValues are stable between expansions
+    assertThat(expansion, equalTo(fromEmpty.expand()));
+
+    List<PCollection<Long>> expectedList =
+        ImmutableList.of(unboundedCount, createOne, boundedCount, maxRecordsCount);
+    for (int i = 0; i < expansion.size(); i++) {
+      assertThat(
+          "Index " + i + " should have equal PValue",
```
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/values/PCollectionListTest.java]: I cannot recall - is this message prepended to the matcher's description, or a replacement? We need the latter, or a more detailed message.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/values/PCollectionListTest.java [hunk 2]
```diff
@@ -44,4 +54,68 @@ public void testEmptyListFailure() {
               + "or must first call empty(Pipeline)"));
     }
   }
+
+  @Test
+  public void testIterationOrder() {
+    Pipeline p = TestPipeline.create();
```
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/values/PCollectionListTest.java]: Narrative comments would improve readability, such as

```
// Build a PCollectionList from a List<PCollection>
...
// Build a PCollectionList via PCollectionList.of(..).and(...)
...
// They should be the equivalent
```

File: sdks/java/core/src/test/java/org/apache/beam/sdk/values/PCollectionListTest.java [hunk 3]
```diff
@@ -44,4 +54,68 @@ public void testEmptyListFailure() {
...
+          "Index " + i + " should have equal PValue",
+          expansion.get(i).getValue(),
+          Matchers.<PValue>equalTo(expectedList.get(i)));
+    }
+  }
+
+  @Test
+  public void testExpansionOrderWithDuplicates() {
+    TestPipeline p = TestPipeline.create();
+    PCollection<Long> firstCount = p.apply("CountFirst", CountingInput.upTo(10L));
```
Comment (kennknowles) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/values/PCollectionListTest.java]: Is there any reason to suspect "duplicates" (which I would expect to mean the _same_ PCollection twice) are relevant? If so, you should definitely save the `PTransform` to a variable and apply it multiple times. That is just a teensy bit "more" duplicate.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/values/PCollectionListTest.java (hunk 1 above)
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/values/PCollectionListTest.java]: Prepended.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/values/PCollectionListTest.java (hunk 3 above)
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/values/PCollectionListTest.java]: Done. I did mean two `==` PCollections, though.

File: sdks/java/core/src/test/java/org/apache/beam/sdk/values/PCollectionListTest.java (hunk 2 above)
Comment (tgroh) [inline sdks/java/core/src/test/java/org/apache/beam/sdk/values/PCollectionListTest.java]: Done.

**Comment by tgroh** [comment]
As a prerequisite to the round-trip elements of the test, I have also added `equals` and `hashCode` methods to `PCollectionList` and `PCollectionTuple`

**Pull request closed without merging**